### ISSUE TITLE
Interaction term on multiple animals issue #13

### ISSUE THREAD

**Issue by windstorm13** [TheoMichelot/moveHMM#15]
Hello

Regarding the use of interaction term (issue #13 analysis for multiple animals and partial pooling) to test for effects of the covariate on different individuals:

I have a dataset of 8 individuals with a total of about 2300 subtracks (about 135700 locations) and I used the formula recommended to bniebuhr

formula = ~ animalID * temp

It's been running for a week now and I have no way of knowing how long it will take to finish.  Can you advise whether it is possible to estimate a run time for training the model, so that I can estimate the remaining run time?

Have there been any updates to moveHMM such as including random effects (I'm not an experienced coder so cannot create custom code for this as McKellar, Towner and DeRuiter did), or if there might be a way to train a moveHMM model in a parallelized computing environment?

Thank you

**Comment by TheoMichelot** [comment]
Hello,

The formula that I suggested in the other thread is actually not really correct, apologies. The animal ID is a categorical covariate and, in moveHMM, it should be included using dummy variables. That is, for K individuals, you need to include K  - 1 binary variables in you data set. Each takes value 1 if the row comes from the corresponding animal, and 0 otherwise. Then, the K - 1 dummy variables should be included as covariates.

In that scenario, the number of estimated parameters increases quickly with the number of individuals, and I suspect that it would not be feasible to fit such a model to time series as long as yours.
Alternatively, you could fit a separate model to each individual. This would also allow you to fit the models in parallel, which is unfortunately not possible within fitHMM at the moment.

Random effects are not implemented in moveHMM, but we are thinking about writing another package for that purpose. Keep an eye out if you're interested!

Best wishes,
Théo

**Comment by bmcclintock** [comment]
Hi all,

The run times are indeed a function of the size of the dataset and the number of parameters that need to be estimated, but also how "good" the starting values are for the optimization.  One option might be to fit separate 'formula = ~ temp' models for each 'animalID' in parallel and then use the fitted values from the individual models as the starting values in the 'formula = ~ animalID * temp' model.  

This approach is demonstrated in the harbour seal example from the momentuHMM vignette (https://github.com/bmcclintock/momentuHMM). momentuHMM is an extension of moveHMM that can be useful for these types of analyses; it can handle categorical covariates such as 'animalID' in the formula argument, and version 1.5.0 (currently only available in the GitHub 'develop' branch) allows for individual-level random effects (sensu McKellar, Towner and DeRuiter) as demonstrated in the pilot whale example of the vignette. In this case, random effects would likely be much more parsimonious than fixed effects (and hopefully reduce run times) while still sufficiently capturing any individual-level variation in the state switching dynamics.

Cheers,
Brett

**Comment by windstorm13** [comment]
Attempting to use momentuHMM for the inclusion of animalID in the formula argument, I have run into an issue - momentuHMM does not accept missing GPS fixes, while moveHMM can accept NA for some missing GPS fixes.  Does this mean that I am then required to use CrawlWrap to 'fill in' those missing fixes prior to using fitHMM with the inclusion of animalID in the formula argument?

**Comment by windstorm13** [comment]
And if so, then using SEX as an random effect, would it be correct to say (formula = ~ ID + SEX) or rather should it be (formula = ~ ID * SEX)?  And if we wanted to incorporate additional covariates of fixed effects such as DISTANCE_WATER would it be correct to say (formula = ~ ID * SEX + DISTANCE_WATER)?

**Comment by windstorm13** [comment]
Attempting to use momentuHMM for the inclusion of animalID in the formula argument, I have run into an issue - momentuHMM does not accept missing GPS fixes, while moveHMM can accept NA for some missing GPS fixes.  Does this mean that I am then required to use CrawlWrap to 'fill in' those missing fixes prior to using fitHMM with the inclusion of animalID in the formula argument?

And if so, then using SEX as an random effect, would it be correct to say (formula = ~ ID + SEX) or rather should it be (formula = ~ ID * SEX)? And if we wanted to incorporate additional covariates of fixed effects such as DISTANCE_WATER would it be correct to say (formula = ~ ID * SEX + DISTANCE_WATER)?

**Comment by bmcclintock** [comment]
moveHMM and momentuHMM handle NAs in data streams (e.g. step length, turn angle) exactly the same way, so I don't understand why this would be an issue for you. However, neither moveHMM nor momentuHMM can handle NAs in covariate values (this is why covariate NAs get filled in by prepData).

The random effects in momentuHMM are discrete-valued (sensu McKellar, Towner, and DeRuiter). These are not specified using the 'formula' argument, they are specified via the 'mixtures' and 'formulaPi' arguments.  Thus is if one wants, say, 2 mixtures with sex effects on the mixture probabilities then one would specify mixtures = 2 and formulaPi = ~SEX. One could also include 'SEX' in the 'formula' argument to include sex effects on both the t.p.m. and mixture probabilities.  See ?fitHMM and the momentuHMM >=1.5.0 vignette for further details.

Just for the record, note that formula = ~ID * SEX is not a good idea.  Sex is an individual-level covariate, and it doesn't make sense to include both ID and any individual-level covariate in the same formula because the resulting model will be overparameterized.

**Comment by HappyHackingOrange** [comment]
I got the exact same issue as windstorm13 when I used the same data I used in moveHMM into momentuHMM it gave the following error due to NA coordinate values: 

![7e8a7cd2-4505-419a-8a80-7028df12be5c](https://user-images.githubusercontent.com/4038198/61743536-59f62700-ad4a-11e9-9065-59f58cf226d8.jpg)

Yet, I did not receive similar error when using moveHMM::prepData.

**Comment by windstorm13** [comment]
I'm sorry, but could you please point me to the pilot whale example of the vignette you referred me to?  I've looked through the vignette and could only see examples for elephant, turtle, northern fulmar, and northern fur, grey, elephant, and harbour seal.  Thank you.

**Comment by bmcclintock** [comment]
@HappyHackingOrange  -- I believe that could only happen if the last observation for a given individual has NA coordinates, in which case perhaps such observations should simply be removed from the data.  Nevertheless, in the '[develop](https://github.com/bmcclintock/momentuHMM/tree/develop)' version of momentuHMM (soon to be released as version 1.5.0), I've added a workaround for when the coordinates of the last observation for an individual are NA. Please let me know if this doesn't solve the issue.

@windstorm13 -- You need momentuHMM >=1.5.0. The vignette for momentuHMM 1.5.0 can currently only be found in the '[develop](https://github.com/bmcclintock/momentuHMM/tree/develop)' branch.

**Comment by HappyHackingOrange** [comment]
Thank you for quick reply.  However, your suggestion did not solve the issue.  I checked to see if there were NA coordinates in all last rows of each individual and there were none.   I also checked first rows as well.

**Comment by bmcclintock** [comment]
Did you successfully install the develop version of momentuHMM?

`devtools::install_github('bmcclintock/momentuHMM@develop')`

If so and you still get the error, can you provide a reproducible example?

**Comment by HappyHackingOrange** [comment]
That's what I did previously, but I figured out the package wasn't installed properly in the first place after admitting to reinstall it.

Also--it's not clear to me how or where the parameters for beta0 are obtained in the pilot whale example. I understand that they are the regression coefficients of the transition probabilities but I do not understand where or how these values were obtained in order to input into beta0 for the fitHMM. As well when specific information to use two mixtures in getPar0, I again do not understand where the additional vector of beta0 values were obtained from? I would appreciate your assistance in clarifying this for me. Thank you.
